lib/excon/socket.rb

```diff
@@ -172,20 +187,21 @@ def connect
       end
     end
 
+    def add_to_read_buffer(str)
+      @read_buffer << str
+      @eof = false
+    end
+
     def read_nonblock(max_length)
       begin
-        if max_length
-          until @read_buffer.length >= max_length
-            @read_buffer << @socket.read_nonblock(max_length - @read_buffer.length)
-          end
-        else
-          loop do
-            @read_buffer << @socket.read_nonblock(@data[:chunk_size])
-          end
+        while !@backend_eof && (!max_length || @read_buffer.length < max_length)
+          @read_buffer << @socket.read_nonblock(@data[:chunk_size])
```

Comment:
I'm not sure how much it matters or not, but this looks like it could read a good deal more than `max_length` into the buffer in some cases. Is that desired? ie I suppose the worst case would be if the buffer contained one less than `max_length` initially. In the old implementation it would just get one byte and end up with `max_length`, whereas with this change it would end up with `2 * max_length - 1`, I think? Does that make sense?

Reply:
I was trying to wrap my head around the new `backend_eof` stuff. I guess this change introduces the need for that as well, if I'm not mistaken? Prior to this change, if there is a `max_length` it should restrict the buffer to that amount, so below when it reads out `max_length` we should be confident that the buffer was emptied. Without that, it's possible that there is more than `max_length` while this is set, causing material to be leftover in the buffer after a read, which this then allows to be read on subsequent calls. Does that match with your understanding/expectations?

I'd like to better understand why we would want to do it this way instead, as it seems a bit counterintuitive to me (and so makes me worry it might introduce problems down the road or be harder to maintain). Does that make sense?

Reply:
Yeah, previously we basically emptied the buffer on ~every call. Which made the buffer not super useful! And causes lots and lots of extra syscalls when we're doing small reads.

Now, we try to read larger chunks of data, basically amortizing the "real work" (syscalls) of calls to read_nonblock. I agree this implementation can be a bit confusing, and would love to think through ways to make it simpler, maybe by abstracting out some of the buffer operations.

But I think fundamentally, we always should be batching reads this way.

Reply:
Yeah, I think that makes sense. Agreed that this seems better in terms of optimization, though it does feel a little harder to follow. This nitty gritty level stuff often seems inherently hard to follow (to me at least). I don't think that should preclude us from these kind of improvements, just that we may want to consider how to keep iterating and refining to make it easier to understand if possible.